File: BeerViewer/Models/Admiral.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using BeerViewer.Models.Raw;

namespace BeerViewer.Models
{
	/// <summary>
	/// 제독 데이터
	/// </summary>
	public class Admiral : RawDataWrapper<kcsapi_basic>
	{
		public string MemberId => this.RawData.api_member_id;
		public string Nickname => this.RawData.api_nickname;

		#region Comment 프로퍼티
		private string _Comment;
		public string Comment
		{
			get { return this._Comment; }
			set
			{
				if (this._Comment != value)
				{
					this._Comment = value;
					this.RaisePropertyChanged();
				}
			}
		}
		#endregion

		/// <summary>
		/// 제독 경험치
		/// </summary>
		public int Experience => this.RawData.api_experience;

		/// <summary>
		/// 다음 레벨까지 남은 경험치
		/// </summary>
		public int ExperienceForNexeLevel => Models.Experience.GetAdmiralExpForNextLevel(this.RawData.api_level, this.RawData.api_experience);

		/// <summary>
		/// 사령부 레벨
		/// </summary>
		public int Level => this.RawData.api_level;

		/// <summary>
		/// 제독 랭크 (원사, 대위, 중위 ...)
		/// </summary>
		public string Rank => Models.Rank.GetName(this.RawData.api_rank);

		/// <summary>
		/// 출격 승리 수
		/// </summary>
		public int SortieWins => this.RawData.api_st_win;

		/// <summary>
		/// 출격 패배 수
		/// </summary>
		public int SortieLoses => this.RawData.api_st_lose;

		/// <summary>
		/// 출격 승률
		/// </summary>
		public double SortieWinningRate
		{
			get
			{
				var battleCount = this.RawData.api_st_win + this.RawData.api_st_lose;
				return battleCount == 0 ? 0
					: this.RawData.api_st_win / (double)battleCount;
			}
		}

		/// <summary>
		/// 소지 가능 칸무스 수
		/// </summary>
		public int MaxShipCount => this.RawData.api_max_chara;

		/// <summary>
		/// 소지 가능 장비 수
		/// </summary>
		public int MaxSlotItemCount => this.RawData.api_max_slotitem + 3;

		internal Admiral(kcsapi_basic RawData) : base(RawData)
		{
			this.Comment = this.RawData.api_comment;
		}

		public override string ToString()
		{
			return $"ID = {this.MemberId}, Nickname = \"{this.Nickname}\", Level = {this.Level}, Rank = \"{this.Rank}\"";
		}
	}
}

File: WindowFrame/src/KanColle/Hub/Homeport/Admiral.ts
import { Component } from "vue-property-decorator";
import Proxy from "@/Proxy/Proxy";
import { ParseKcsApi } from "@KC/Functions";
import KanColleStoreClient, { StoreInterface } from "@KC/Store/KanColleStoreClient";

import { kcsapi_port } from "@KC/Raw/kcsapi_port";
import { kcsapi_member_updatecomment } from "@KC/Raw/kcsapi_member";
import { kcsapi_require_info } from "@KC/Raw/kcsapi_require_info";
import { kcsapi_basic } from "@KC/Raw/kcsapi_basic";

@Component
export default class Admiral extends KanColleStoreClient {
	constructor() {
		super();

		Proxy.Instance.Register("/kcsapi/api_get_member/require_info", (req, resp) => {
			ParseKcsApi<kcsapi_require_info>(resp, (x) => this.Update(x.api_basic));
		});
		Proxy.Instance.Register("/kcsapi/api_get_member/basic", (req, resp) => {
			ParseKcsApi<kcsapi_basic>(resp, (x) => this.Update(x));
		});
		Proxy.Instance.Register("/kcsapi/api_port/port", (req, resp) => {
			ParseKcsApi<kcsapi_port>(resp, (x) => this.Update(x.api_basic));
		});
		Proxy.Instance.Register("/kcsapi/api_req_member/updatecomment", (req, resp) => {
			ParseKcsApi<unknown, kcsapi_member_updatecomment>(resp, req.body, (x, y) => {
				this.StoreAdmiralCommentUpdate(y.api_cmt);
			});
		});
	}

	private Update(data: kcsapi_basic) {
		this.StoreAdmiralUpdate({
			MemberId: parseInt(data.api_member_id, 10),
			Rank: data.api_rank,

			Level: data.api_level,
			Experience: data.api_experience,

			Nickname: data.api_nickname,
			Comment: data.api_comment,

			Sortie: {
				Wins: data.api_st_win,
				Loses: data.api_st_lose,
			},

			Maximum: {
				Ships: data.api_max_chara,
				Equips: data.api_max_slotitem,
			},
		});
	}
}
